### ISSUE TITLE
Not listed as "media player" application

### ISSUE THREAD

**Issue by cziss** [TJYSunset/Phocid#12]
Hi, I've noticed that player is not recognized as "media player" application. So you can not make media routines etc. I have Android 14, Samsung with One UI 6.1.

**Comment by TJYSunset** [comment]
Can you provide an example of any open source media player that supports said feature? I can find absolutely zero information on how to integrate with Bixby, nor do I own a Samsung device.

Alternatively, which operations can Bixby perform on media players? I suspect Bixby just uses standard Android intents, but Android's docs are bad at listing everything available.

**Comment by cziss** [comment]
I don't meant Bixby. Just routines similar on plain Android. Also player is not listed in Android auto (even also if "unknown sources" is setted up in developers menu). I've used Metro player was listed as "media player" in routines and also in Android auto.

Routines can set different equaliser or Dolby Atmos etc. when media player playing.

**Comment by TJYSunset** [comment]
I've supposedly marked the app as a music player, but implemented non of the intents (because they either look irrelevant or have zero documentation). How does this look like in your use case?

https://github.com/user-attachments/files/17994718/app-release.zip

**Comment by cziss** [comment]
Unfortunately it's still not listed as "media player" in routines or Android Auto.

**Comment by TJYSunset** [comment]
Could you try this build?
https://github.com/TJYSunset/Phocid/releases/tag/v20241204

Sorry, but there's no information on the entire internet about this topic, and I couldn't find what you've mentioned in the official Android Auto emulator. (The emulator refuses to display any media files at all, the music files I uploaded disappears after reboot, and Metro Player straight up crashes after a splash screen.)

**Comment by cziss** [comment]
No, it also not helped. Ok, understand - I also did some internet research, but without any useful information. Strange is, that audiobook player Voice, VLC, or NewPipe somehow managed to accomplish to be listed as media app. Even Libera FD or Total Commander, which plays media marginally.

**Comment by TJYSunset** [comment]
Some more random attempts:

Simply marking service as exported:
[app-release.zip](https://github.com/user-attachments/files/18009355/app-release.zip)

Adding even more intents:
[app-release.zip](https://github.com/user-attachments/files/18009410/app-release.zip)

Do you have other apps that **can utilize** media players similar to routines? I just can't get this Android Auto emulator running (looks like Google really doesn't want to have any Android Auto developers.)

**Comment by cziss** [comment]
I've tried it, but still no success. Yes, that I've mentioned above, all are on F-Droid. Well I thing, that if app is "media player" it's enough to be listed in Android Auto (when you have unknown sources ON). At least for Voice or Metro player.

**Comment by TJYSunset** [comment]
I meant any other apps that can scan what media players you have installed, so I can test it myself.

**Comment by cziss** [comment]
Sorry. No, I don't know other apps. I think, that it most by done somehow by Android itself.

**Comment by cziss** [comment]
Hi, I've tried to ask how to solved it in Voice app, [here](https://github.com/PaulWoitaschek/Voice/discussions/2641).

**Comment by ulipo** [comment]
I've noticed the same issue with [Bluetooth connect & play](https://play.google.com/store/apps/details?id=com.cp2.start.and.play.music.player&hl=it) app:
Phocid is not listed in media players, it appears if I check "all apps" in configuration but still won't play at bluetooth connection (it works with Phonograph, tho).

**Comment by TJYSunset** [comment]
[v20241221](https://github.com/TJYSunset/Phocid/releases/tag/v20241221) should now be listed in the "Bluetooth connect & play" app as a music player, but will not start playing automatically like Phonograph does because of reasons* (LineageOS' built-in music app doesn't anyway).

\*reasons: Phonograph seems to have a self-contained service that starts when it receives an intent from "Bluetooth connect & play". Phocid has complex additional logic tied to its UI because of more reasons\*\*, so it has to start an activity instead. [Android disallows starting activities from background.](https://developer.android.com/guide/components/activities/background-starts)

\*\*more reasons: Google made fancy new frameworks that you "should" use, which lack absolutely every function (like getting and setting the shuffle order). Phocid uses twisted hacks to regain lacking functions, violating the frameworks' interop contracts.

**Comment by ulipo** [comment]
Damn, Google is making very difficult to _own_ a smartphone!
I can confirm it is listed in th Bluetooth app now, thanks a lot!

**Comment by lineinthesand** [comment]
I also noticed that Phocid is not recognized as a music player for some reason in the last.fm app (I saw the other issue but I think this one is actually the main one) despite it being announced as such in the manifest.
But could you please take a look at this manifest (Auxio music player) https://github.com/OxygenCobalt/Auxio/blob/dev/app/src/main/AndroidManifest.xml

It has an intent-filter section with an interesting comment:
```
Signal that we can take audio of *any* format, file or content. This is important, as some apps use this to determine if Auxio is eligible to be a default music player.
```
Maybe this could solve the problem of Phocid currently not being recognized as a media player.

**Comment by TJYSunset** [comment]
Phocid can't actually open arbitrary audio sources (it's designed to only work with MediaStore entries - maybe can be expanded to work with other persistent sources, but not temporary sources like most intent senders), so that would be a new defect if someone opens something with Phocid and meets an error message. As long as you can force Last.fm to use Phocid, I think that workaround is not worth it.

**Comment by lineinthesand** [comment]
I'll try if I can find any sources in that last.fm app and if there are, I'll try to check whether something can be changed for the mechanism on how it recognizes media apps.

**Comment by Catfriend1** [comment]
I think this is not a separate issue and related to this ticket here.

I am very happy Phocid supports the folder structure and looks like the same as if I use my file manager app (Material Files). But I cannot open a single mp3 file from Material files. The path lies normally on my internal storage /storage/emulated0/Music/.... Can that be implemented ? Other apps that are not compatible with the folder structure play them fine from there, for example App Manager or Cute Music (both taken from Fdroid). I am on Android 15.

**Comment by TJYSunset** [comment]
@Catfriend1 It's easy to implement, but what should be presented to the user if the intent URI is `/no-permission-here/file.mp3` or `https://example.com/file.mp3`?
---
Tested with some random apps and they present... absolutely nothing. Not even a error message. OK, maybe it's not so bad to just display an error toast.

**Comment by Catfriend1** [comment]
@TJYSunset yes, I totally understand your concern. No designer likes to display something hardly understandable to the user which is non tech savy like "hey, Android SAF restricts me so your /Downloads/1.mp3 will not play because of OS restrictions. Do it in app and get away with not understanding why it is so weird". I have similar problems with "my" Syncthing Android users. How can I explain them in easy and short to understand form why the Downloads folder is non-selectable from SAF folder picker? And that only on Android 10+? Or why they can use sdcard's in read only mode pre-Android 11 while internal storage folder are writable to the SyncthingNative process (cross compiled exe). It's ugly, and I understand if you'd keep clear instead. Else I'd suggest to display the truth: "Sorry, we can't play this file because of Android OS restrictions that apply to the parent folder. Move it to (e.g. music folder) or listen from in-app."

**Comment by TJYSunset** [comment]
Phocid can now handle `ACTION_VIEW` intent in [v20250521](https://github.com/TJYSunset/Phocid/releases/tag/v20250521).

---

Rant: I thought I could just parse the intent URI to obtain a reasonable file path, because that's what my file manager sends... Then I got `/file%3A%2F%2F%2Fstorage%2Femulated%2F0%2FMusic%2FEXAMPLE%2520-%2520EXAMPLE%2520(EXAMPLE).mp3` from this Material Files app (already URL decoded). Ah yes, the fabled triple URL encoding.

I ended up having to shamelessly copy other player's source code.

**Comment by Catfriend1** [comment]
Thank you very much! 🙂

**Comment by TJYSunset** [comment]
The original post is about Samsung - we still not know if that's resolved unless a Samsung user confirms.

I suspect it might also have something to do with not being installed from Google Play - [Android Auto won't display apps from "unknown sources" unless using Developer Options](https://www.reddit.com/r/AndroidAuto/comments/bxicf1/music_apps_not_showing_in_android_auto/), for example.

**Comment by MateusRodCosta** [comment]
Hi, I checked latest version on F-Droid (v20250821) on a OneUI 7 device and it doesn't show up as a media player.

If it's any useful I have a audio app which does show up, here is its manifest:

https://github.com/MateusRodCosta/vidya_music/blob/main/android/app/src/main/AndroidManifest.xml

**Comment by MateusRodCosta** [comment]
So, when running [v20250823](https://github.com/TJYSunset/Phocid/releases/tag/v20250823), I tested play/pause and restoring the media player (closed the app from overview and dismissed the notification) with both Spotify and Phocid and both seem to work fine.

So, yeah, probably there's not that much need of actually implementing the MediaButtonReceiver over the dummy (or more likely your current implementation is enough).

If it's any useful, the Flutter app I sent earlier uses the [audio_service package](https://pub.dev/packages/audio_service) underneath which has the following definition: https://github.com/ryanheise/audio_service/blob/minor/audio_service/android/src/main/java/com/ryanheise/audioservice/MediaButtonReceiver.java

It would seem to me though that the appropriate way is really using media3 MediaButtonReceiver or a class based on that (if needed), which is the latest apk you provided.

The latest apk you sent which uses `androidx.media3.session.MediaButtonReceiver` seems to behave exactly the same while also using Google's own handling of the headphone button.
I did not manage to trigger the toast you mentioned by just closing the app via the Overview and dismissing the notification and then triggering an app re-open via the button.

In any case, I don't use the headphone that much but it might be a useful feature for other users, so for me it's fine if you track this in a new issue or postpone handling this.
I would recommend implementing as the Android documentation says, but if the docs just don't match reality (normally I would assume OEM meddling with Android internals, but you mentioned Lineage) then I don't know what to suggest =/.

Still thanks for taking time to look into this. I think that's all from my side.
